src/python/grpcio_tests/tests/observability/_open_telemetry_observability_test.py

```diff
@@ -0,0 +1,232 @@
+
+    def testRecordStreamUnary(self):
+        otel_plugin = BaseTestOpenTelemetryPlugin(self._provider)
+
+        with grpc_observability.OpenTelemetryObservability(
+            plugins=[otel_plugin]
+        ):
+            server, port = _test_server.start_server()
+            self._server = server
+            _test_server.stream_unary_call(port=port)
+
+        self._validate_metrics_exist(self.all_metrics)
+
+    def testRecordStreamStream(self):
+        otel_plugin = BaseTestOpenTelemetryPlugin(self._provider)
+
+        with grpc_observability.OpenTelemetryObservability(
+            plugins=[otel_plugin]
+        ):
+            server, port = _test_server.start_server()
+            self._server = server
+            _test_server.stream_stream_call(port=port)
+
+        self._validate_metrics_exist(self.all_metrics)
+
+    def testTargetAttributeFilter(self):
+        # If target label contains 'localhost', is should be replaced with 'other'.
+        def target_filter(target: str) -> bool:
+            if "localhost" in target:
+                return False
+            return True
+
+        otel_plugin = BaseTestOpenTelemetryPlugin(self._provider)
+        otel_plugin.target_attribute_filter = target_filter
+
+        with grpc_observability.OpenTelemetryObservability(
+            plugins=[otel_plugin]
+        ):
+            server, port = _test_server.start_server()
+            self._server = server
+            _test_server.unary_unary_call(port=port)
+
+        self._validate_metrics_exist(self.all_metrics)
+        target_values = self.all_metrics["label"].get(GRPC_TARGET_LABEL)
+        self.assertTrue(GRPC_OTHER_LABEL_VALUE in target_values)
+
+    def testMethodAttributeFilter(self):
+        # If method label contains 'UnaryUnaryFiltered', is should be replaced with 'other'.
+        FILTERED_METHOD_NAME = "UnaryUnaryFiltered"
+
+        def method_filter(method: str) -> bool:
+            if FILTERED_METHOD_NAME in method:
+                return False
+            return True
+
+        otel_plugin = BaseTestOpenTelemetryPlugin(self._provider)
+        otel_plugin.generic_method_attribute_filter = method_filter
+
+        with grpc_observability.OpenTelemetryObservability(
+            plugins=[otel_plugin]
+        ):
+            server, port = _test_server.start_server()
+            self._server = server
+            _test_server.unary_unary_call(port=port)
+            _test_server.unary_unary_filtered_call(port=port)
+
+        self._validate_metrics_exist(self.all_metrics)
+        method_values = self.all_metrics["label"].get(GRPC_METHOD_LABEL)
+        self.assertTrue(GRPC_OTHER_LABEL_VALUE in method_values)
+        self.assertTrue(FILTERED_METHOD_NAME not in method_values)
+
+    def _validate_metrics_exist(self, all_metrics: Dict[str, Any]) -> None:
+        # Sleep here to make sure we have at least one export from OTel MetricExporter.
+        time.sleep(OTEL_EXPORT_INTERVAL_S)
```

Comment:
In general, I'm very wary of using statically timed sleeps in a test. They tend to produce flakiness because, even if it can be proven that two threads of execution _should_ interleave a certain way because of sleeps with specific times they have in their implementation, thread pre-emption creates cases where they don't _always_ interleave the way intended. I strongly suggest that you make this assertion _eventual_. In other words, retry multiple times up to a configurable (long -- multiple seconds) timeout. Check out this [`assert_eventually`](https://github.com/grpc/grpc/blob/6d35000c17d96819184f344183b07104f5b7eb16/src/python/grpcio_tests/tests_py3_only/unit/_simple_stubs_test.py#L170) higher order function we already have for cases like this.

We should also characterize how flaky this test is already by running under bazel 1000 times. We should aim for fewer than 2 failures/timeouts out of 1000.